### ISSUE TITLE
Crash when deconstructing an entity when last fuel is consumed.

### ISSUE THREAD

**Issue by gonsen71** [GDQuest/godot-2d-builder#9]
I've found a bug when deconstructing the `StirlingEngineEntity` and the `FurnaceEntity` when the last fuel is consumed, no need to be fully consumed. For example, if you put a single `Lumber` in the `StirlingEngineEntity`, this `Lumber` is deleted but the engine will work until is fully consumed. If you then deconstruct the item, the game will crash, if you let it to be fully consumed and then deconstruct the item, the game will crash too:

```
Invalid call. Nonexistent function 'get_parent' in base 'Nil'.
0 - res://Systems/EntityPlacer.gd:207 - at function: _drop_entity
1 - res://Systems/EntityPlacer.gd:275 - at function: _finish_deconstruct
```

An easy fix would be to wrap `panel.held_item` with `is_instance_valid` in `get_inventory() `function of `InventoryBar.gd`:

```gdscript
func get_inventory() -> Array:
	var output := []
	for panel in panels:
		if is_instance_valid(panel.held_item):
			output.push_back(panel.held_item)

	return output
```

But then I thought why the held_item is not null in the first instance, I started looking more deeply, and I found that the fuel is freed in the entities but in the panel is not set to null.

```gdscript
# StirlingEngineEntity.gd & FurnaceEntity.gd
# ...
func _consume_fuel(amount: float) -> void:
	# ...
		if gui.gui.fuel.stack_count == 0:
			gui.gui.fuel.queue_free()
			gui.gui.fuel = null
	# ...
```

After analyzing and thinking what would be the best solution I thought that adding `is_instance_valid` everywhere feels wrong to me. And then I thought that using the GUI to store and manage the actual Blueprints and inventory it felt wrong to me also, from the very beginning, the GUI is doing too much.  
Wouldn't be better to decouple the inventory system that manages the items and inventories and let the GUI care **only** about showing and displaying that data?  
The inventory could have a function to delete an item from an inventory, free it from the memory and then emit a signal so others could nullify the variables.

Anyways, I've already overthought it too much. I've made a pull request with the easy fix: [Fix deconstruct when last fuel is consumed](https://github.com/GDQuest/godot-2d-builder/pull/8)

**Comment by Razoric480** [comment]
The codebase could very well be further broken down and decoupled, but I had to stop somewhere for the sake of making a course out of it. Time and simplicity had to win out.

If I was just developing it as a game alone I'd have kept digging deeper to find and avoid these kind of issues.

Thanks for the simple fix, though, will pull it in and update the course accordingly.

**Comment by Razoric480** [comment]
Fixed by #8

**Comment by gonsen71** [comment]
Yes, makes sense. Thank you for this amazing course!
